plenum/server/replica.py

```diff
@@ -819,27 +824,73 @@ def trackBatches(self, pp: PrePrepare, prevStateRootHash):
-                r += 1
-
-        if r > 0:
-            self.lastBatchCreated = time.perf_counter()
-        return r
+    def send_3pc_batch(self):
+        sent_batches = set()
+
+        # 1. send 3PC batches with requests for every ledger
+        self._send_3pc_batches_for_ledgers(sent_batches)
+
+        # 2. for every ledger we havne't just sent a 3PC batch check if it's not fresh enough,
+        # and send an empty 3PC batch to update the state if needed
+        self._send_3pc_freshness_batch(sent_batches)
+
+        # 3. update ts of last sent 3PC batch
+        sent_batches.discard(None)
+        if len(sent_batches) > 0:
+            self.lastBatchCreated = self.get_current_time()
+
+        return len(sent_batches)
+
+    def _send_3pc_batches_for_ledgers(self, sent_batches):
+        for ledger_id, q in self.requestQueues.items():
+            if len(q) == 0:
+                continue
+
+            queue_full = len(q) >= self.config.Max3PCBatchSize
+            timeout = self.lastBatchCreated + self.config.Max3PCBatchWait < self.get_current_time()
+            if not queue_full and not timeout:
+                continue
+
+            sent_batches.add(
+                self._do_send_3pc_batch(ledger_id=ledger_id, is_empty=False))
+
+    def _send_3pc_freshness_batch(self, sent_batches):
+        if not self.config.UPDATE_STATE_FRESHNESS:
+            return
+
+        if not self.isMaster:
+            return
+
+        self._freshness_checker.check_freshness(self.get_current_time())
+
+        # Update freshness for all outdated ledgers sequentially without any waits
+        # Consider sending every next update in Max3PCBatchWait only
+        while self._freshness_checker.get_outdated_ledgers_count() > 0:
+            ledger_id, ts = self._freshness_checker.pop_next_outdated_ledger()
+            if ledger_id in sent_batches:
+                self.logger.debug("Ledger {} is not updated for {} seconds, "
+                                  "but a 3PC for this ledger has been just sent".format(ledger_id, ts))
+                continue
+
+            self.logger.info("Ledger {} is not updated for {} seconds, "
+                             "so its freshness state is going to be updated now".format(ledger_id, ts))
+            sent_batches.add(
+                self._do_send_3pc_batch(ledger_id=ledger_id, is_empty=True))
+
+    def _do_send_3pc_batch(self, ledger_id, is_empty=False):
+        oldStateRootHash = self.stateRootHash(ledger_id, to_str=False)
+        pre_prepare = self.create_3pc_batch(ledger_id, is_empty)
+        if pre_prepare is None:
+            return None
+        self.sendPrePrepare(pre_prepare)
+        if not self.isMaster:
+            self.node.last_sent_pp_store_helper.store_last_sent_pp_seq_no(
+                self.instId, pre_prepare.ppSeqNo)
+        self.trackBatches(pre_prepare, oldStateRootHash)
+        return ledger_id
+
+    def get_current_time(self):
+        return time.perf_counter()
```

Comment:
Did I get it right that this is done solely to be able to inject time in tests? Great to have this, however I used to do it in other way:
```
class ItShouldBeTestable:
    def __init__(self, perf_counter=time.perf_counter):
        self.perf_counter = perf_counter

    def doSomething():
        x = self.perf_counter()

...

def test_this():
    ts = MockTimestamp()
    x = ItShouldBeTestable(perf_counter=ts)
    // do whatever we want
```
In my opinion this better shows intent, however this is also subjective.

By the way, we already have `MockTimestamp` class in common test helpers which is used in several tests.

Reply:
Yes, this is done just for tests.